osquery/carver/carver.cpp

```diff
@@ -0,0 +1,304 @@
+
+/*
+ * Carver block chunking size. This specifies how big of a block size to
+ * use when carving a file from disk.
+ */
+const size_t kBuffSize = 8192;
+
+DECLARE_string(tls_hostname);
+
+/// Session creation endpoint for forensic file carve
+CLI_FLAG(string,
+         carver_start_endpoint,
+         "",
+         "TLS/HTTPS init endpoint for forensic carver");
+
+/// Data aggregation endpoint for forensic file carve
+CLI_FLAG(
+    string,
+    carver_continue_endpoint,
+    "",
+    "TLS/HTTPS endpoint that receives carved content after session creation");
+
+/// Size of blocks used for POSTing data back to remote endpoints
+CLI_FLAG(int32,
+         carver_block_size,
+         8192,
+         "Size of blocks used for POSTing data back to remote endpoints");
+
+CLI_FLAG(bool,
+         disable_carver,
+         true,
+         "Disable the osquery file carver (default true)");
+
+/// Helper function to update values related to a carve
+void updateCarveValue(const std::string& guid,
+                      const std::string& key,
+                      const std::string& value) {
+  std::string carve;
+  auto s = getDatabaseValue(kQueries, kCarverDBPrefix + guid, carve);
+  if (!s.ok()) {
+    VLOG(1) << "Failed to update status of carve in database " << guid;
+    return;
+  }
+
+  pt::ptree tree;
+  try {
+    std::stringstream ss(carve);
+    pt::read_json(ss, tree);
+  } catch (const pt::ptree_error& e) {
+    VLOG(1) << "Failed to parse carve entries: " << e.what();
+    return;
+  }
+
+  tree.put(key, value);
+
+  std::ostringstream os;
+  pt::write_json(os, tree, false);
+  s = setDatabaseValue(kQueries, kCarverDBPrefix + guid, os.str());
+  if (!s.ok()) {
+    VLOG(1) << "Failed to update status of carve in database " << guid;
+    return;
+  }
+}
+
+Carver::Carver(const std::set<std::string>& paths, const std::string& guid) {
+  status_ = Status(0, "Ok");
+  for (const auto& p : paths) {
+    carvePaths_.insert(fs::path(p));
+  }
+
+  // Construct the uri we post our data back to:
+  startUri_ = TLSRequestHelper::makeURI(FLAGS_carver_start_endpoint);
+  contUri_ = TLSRequestHelper::makeURI(FLAGS_carver_continue_endpoint);
+
+  // Generate a unique identifier for this carve
+  carveGuid_ = guid;
+
+  // TODO: Adding in a manifest file of all carved files might be nice.
+  carveDir_ =
+      fs::temp_directory_path() / fs::path(kCarvePathPrefix + carveGuid_);
+  auto ret = fs::create_directory(carveDir_);
+  if (!ret) {
+    status_ = Status(1, "Failed to create carve file store");
+  }
+
+  // Store the path to our archive for later exfiltration
+  archivePath_ = carveDir_ / fs::path(kCarveNamePrefix + carveGuid_ + ".tgz");
+
+  // Update the DB to reflect that the carve is pending.
+  updateCarveValue(carveGuid_, "status", "PENDING");
+};
+
+Carver::~Carver() {
+  fs::remove_all(carveDir_);
+}
+
+void Carver::start() {
+  // If status_ is not Ok, the creation of our tmp FS failed
+  if (!status_.ok()) {
+    return;
```

Comment:
I'll add more logging to it, but `start` is what gets called by the dispatcher, so there isn't a great place anywhere to check a return value. I could drop a warning log here? As if this hits it's likely we really don't want to continue with the carve.